ibis/backends/dask/execution/generic.py

```diff
@@ -0,0 +1,869 @@
+"""Execution rules for generic ibis operations."""
+
+import collections
+import datetime
+import decimal
+import functools
+import numbers
+import operator
+from typing import Optional
+
+import dask.array as da
+import dask.dataframe as dd
+import numpy as np
+import pandas as pd
+from dask.dataframe.groupby import DataFrameGroupBy, SeriesGroupBy
+from pandas import isnull, to_datetime
+from pandas.api.types import DatetimeTZDtype
+
+import ibis.common.exceptions as com
+import ibis.expr.datatypes as dt
+import ibis.expr.operations as ops
+import ibis.expr.types as ir
+from ibis.backends.pandas.core import (
+    boolean_types,
+    execute,
+    floating_types,
+    integer_types,
+    numeric_types,
+    scalar_types,
+    simple_types,
+)
+from ibis.backends.pandas.execution import constants
+from ibis.backends.pandas.execution.generic import execute_node
+from ibis.expr.scope import Scope
+from ibis.expr.timecontext import TIME_COL
+from ibis.expr.typing import TimeContext
+
+# from .. import aggcontext as agg_ctx
+from ..client import DaskClient, DaskTable
+
+
+@execute_node.register(ops.Limit, dd.DataFrame, integer_types, integer_types)
+def execute_limit_frame(op, data, nrows, offset, **kwargs):
+    return data.loc[offset : offset + nrows]
+
+
+@execute_node.register(ops.Cast, SeriesGroupBy, dt.DataType)
+def execute_cast_series_group_by(op, data, type, **kwargs):
+    result = execute_cast_series_generic(op, data.obj, type, **kwargs)
+    return result.groupby(data.grouper.groupings)
+
+
+@execute_node.register(ops.Cast, dd.Series, dt.DataType)
+def execute_cast_series_generic(op, data, type, **kwargs):
+    return data.astype(constants.IBIS_TYPE_TO_PANDAS_TYPE[type])
+
+
+@execute_node.register(ops.Cast, dd.Series, dt.Array)
+def execute_cast_series_array(op, data, type, **kwargs):
+    value_type = type.value_type
+    numpy_type = constants.IBIS_TYPE_TO_PANDAS_TYPE.get(value_type, None)
+    if numpy_type is None:
+        raise ValueError(
+            'Array value type must be a primitive type '
+            '(e.g., number, string, or timestamp)'
+        )
+    return data.map(
+        lambda array, numpy_type=numpy_type: list(map(numpy_type, array))
+    )
+
+
+@execute_node.register(ops.Cast, dd.Series, dt.Timestamp)
+def execute_cast_series_timestamp(op, data, type, **kwargs):
+    arg = op.arg
+    from_type = arg.type()
+
+    if from_type.equals(type):  # noop cast
+        return data
+
+    tz = type.timezone
+
+    if isinstance(from_type, (dt.Timestamp, dt.Date)):
+        return data.astype(
+            'M8[ns]' if tz is None else DatetimeTZDtype('ns', tz)
+        )
+
+    if isinstance(from_type, (dt.String, dt.Integer)):
+        timestamps = data.map_partitions(
+            to_datetime,
+            infer_datetime_format=True,
+            meta=(data.name, 'datetime64[ns]'),
```

Comment:
Do we need to include timezone in the type?

Reply:
I believe that since we're casting from str/int `pd.infer_datetime` will properly grab the tz info as in https://github.com/ibis-project/ibis/blob/bf1196624f983dea46a9c7bafd814e27fdc56577/ibis%2Fbackends%2Fdask%2Ftests%2Fexecution%2Ftest_temporal.py#L84.

Reply:
yeah this is prb ok

Reply:
Sorry I don't totally understand why this is OK. It looks to me that this is telling Dask that the type of dd.Series is "datetime64[ns]", but the actual type of the output is "datetime64[ns, tz]"?